handlers/dashboard.go

```diff
@@ -28,36 +30,128 @@ import (
 	if str == "" {
-		return []uint64{}, nil
+		return []uint64{}, [][]byte{}, nil
 	}
 
 	strSplit := strings.Split(str, ",")
 	strSplitLen := len(strSplit)
 
 	// we only support up to [validatorLimit] validators
 	if strSplitLen > validatorLimit {
-		return []uint64{}, ErrTooManyValidators
+		return []uint64{}, [][]byte{}, ErrTooManyValidators
 	}
 
-	validators := make([]uint64, strSplitLen)
-	keys := make(map[uint64]bool, strSplitLen)
+	var validatorIndices []uint64
+	var validatorPubkeys [][]byte
+	keys := make(map[interface{}]bool, strSplitLen)
 
-	for i, vStr := range strSplit {
+	// Find all pubkeys
+	for _, vStr := range strSplit {
+		if !searchPubkeyExactRE.MatchString(vStr) {
+			continue
+		}
+		if !strings.HasPrefix(vStr, "0x") {
+			// Query string public keys have to have 0x prefix
+			return []uint64{}, [][]byte{}, fmt.Errorf("invalid pubkey")
+		}
+		// make sure keys are unique
+		if exists := keys[vStr]; exists {
+			continue
+		}
+		keys[vStr] = true
+		validatorPubkeys = append(validatorPubkeys, common.FromHex(vStr))
+
+	}
+
+	// Convert pubkeys to indices if possible
+	if len(validatorPubkeys) > 0 {
+		validatorInfos := []struct {
+			Index  uint64
+			Pubkey []byte
+		}{}
+		err := db.ReaderDb.Select(&validatorInfos, `SELECT validatorindex as index, pubkey FROM validators WHERE pubkey = ANY($1)`, validatorPubkeys)
+		if err != nil {
+			return []uint64{}, [][]byte{}, err
+		}
+
+		for _, info := range validatorInfos {
+			validatorIndices = append(validatorIndices, info.Index)
+			keys[info.Index] = true
+			for idx, pubkey := range validatorPubkeys {
+				if bytes.Contains(pubkey, info.Pubkey) {
+					validatorPubkeys = append(validatorPubkeys[:idx], validatorPubkeys[idx+1:]...)
+					break
+				}
+			}
+		}
+	}
+
+	// Find all indices
+	for _, vStr := range strSplit {
+		if searchPubkeyExactRE.MatchString(vStr) {
+			continue
+		}
 		v, err := strconv.ParseUint(vStr, 10, 64)
 		if err != nil {
-			return []uint64{}, err
+			return []uint64{}, [][]byte{}, err
 		}
-		// make sure keys are uniq
+		// make sure keys are unique
 		if exists := keys[v]; exists {
 			continue
 		}
 		keys[v] = true
-		validators[i] = v
+		validatorIndices = append(validatorIndices, v)
+	}
+
+	return validatorIndices, validatorPubkeys, nil
+}
+
+func updateValidatorsQueryString(w http.ResponseWriter, r *http.Request, validatorIndices []uint64, validatorPubkeys [][]byte) {
+	validatorsCount := len(validatorIndices) + len(validatorPubkeys)
+	if validatorsCount == 0 {
+		return
+	}
+	strValidators := make([]string, validatorsCount)
+	for i, n := range validatorIndices {
+		strValidators[i] = fmt.Sprintf("%v", n)
+	}
+	for i, n := range validatorPubkeys {
+		strValidators[i+len(validatorIndices)] = fmt.Sprintf("%#x", n)
+	}
+
+	q := r.URL.Query()
+	q.Set("validators", strings.Join(strValidators, ","))
+	r.URL.RawQuery = q.Encode()
+
+	fmt.Fprintf(w, `<script>window.history.replaceState(null, "Dashboard", "%s");</script>`, r.URL.String())
```

Comment:
I thought so too, it would be the only line in the code that calls a script.
I used it because it was the only way I found to change the URL without reloading the page which leads to the `parseValidatorsFromQueryString` calculation being done twice.

Since I think redirect works better with the other code parts where we now want to use updateValidatorsQueryString I used it.